File: server/hotrod/src/test/scala/org/infinispan/server/hotrod/HotRod10ReplicationTest.scala
package org.infinispan.server.hotrod

import java.lang.reflect.Method
import org.infinispan.configuration.cache.{CacheMode, ConfigurationBuilder}
import org.infinispan.server.hotrod.OperationStatus._
import org.infinispan.test.AbstractCacheTest._
import org.infinispan.test.TestingUtil
import org.testng.Assert._
import org.testng.annotations.Test
import test.AbstractTestTopologyAwareResponse
import test.HotRodTestingUtil._

/**
 * Tests Hot Rod instances configured with replication in protocol version 1.0.
 *
 * @author Galder Zamarreño
 * @since 4.1
 */
@Test(groups = Array("functional"), testName = "server.hotrod.HotRodReplicationTest")
class HotRod10ReplicationTest extends HotRodMultiNodeTest {

   override protected def cacheName: String = "hotRodReplSync"

   override protected def createCacheConfig: ConfigurationBuilder = {
      val config = hotRodCacheConfiguration(
         getDefaultClusteredCacheConfig(CacheMode.REPL_SYNC, false))
      config.clustering().stateTransfer().fetchInMemoryState(true)
      config
   }

   override protected def protocolVersion = 10

   def testReplicatedPut(m: Method) {
      val resp = clients.head.put(k(m) , 0, 0, v(m))
      assertStatus(resp, Success)
      assertSuccess(clients.tail.head.get(k(m), 0), v(m))
   }

   def testReplicatedPutIfAbsent(m: Method) {
      assertKeyDoesNotExist(clients.head.assertGet(m))
      assertKeyDoesNotExist(clients.tail.head.assertGet(m))
      val resp = clients.head.putIfAbsent(k(m) , 0, 0, v(m))
      assertStatus(resp, Success)
      assertSuccess(clients.tail.head.get(k(m), 0), v(m))
      assertStatus(clients.tail.head.putIfAbsent(k(m) , 0, 0, v(m, "v2-")), OperationNotExecuted)
   }

   def testReplicatedReplace(m: Method) {
      var resp = clients.head.replace(k(m), 0, 0, v(m))
      assertStatus(resp, OperationNotExecuted)
      resp = clients.tail.head.replace(k(m), 0, 0, v(m))
      assertStatus(resp , OperationNotExecuted)
      clients.tail.head.assertPut(m)
      resp = clients.tail.head.replace(k(m), 0, 0, v(m, "v1-"))
      assertStatus(resp, Success)
      assertSuccess(clients.head.assertGet(m), v(m, "v1-"))
      resp = clients.head.replace(k(m), 0, 0, v(m, "v2-"))
      assertStatus(resp, Success)
      assertSuccess(clients.tail.head.assertGet(m), v(m, "v2-"))
   }

   def testPingWithTopologyAwareClient() {
      var resp = clients.head.ping
      assertStatus(resp, Success)
      assertEquals(resp.topologyResponse, None)

      resp = clients.tail.head.ping(1, 0)
      assertStatus(resp, Success)
      assertEquals(resp.topologyResponse, None)

      resp = clients.head.ping(2, 0)
      assertStatus(resp, Success)
      assertTopologyReceived(resp.topologyResponse.get, servers, currentServerTopologyId)

      resp = clients.tail.head.ping(2, 0)
      assertStatus(resp, Success)
      assertTopologyReceived(resp.topologyResponse.get, servers, currentServerTopologyId)

      resp = clients.tail.head.ping(2, 3)
      assertStatus(resp, Success)
      assertEquals(resp.topologyResponse, None)
   }

   def testReplicatedPutWithTopologyChanges(m: Method) {
      var resp = clients.head.put(k(m) , 0, 0, v(m), 1, 0)
      assertStatus(resp, Success)
      assertEquals(resp.topologyResponse, None)
      assertSuccess(clients.tail.head.get(k(m), 0), v(m))

      resp = clients.head.put(k(m) , 0, 0, v(m, "v1-"), 2, 0)
      assertStatus(resp, Success)
      assertTopologyReceived(resp.topologyResponse.get, servers, currentServerTopologyId)

      resp = clients.tail.head.put(k(m) , 0, 0, v(m, "v2-"), 2, 0)
      assertStatus(resp, Success)
      assertTopologyReceived(resp.topologyResponse.get, servers, currentServerTopologyId)

      resp = clients.head.put(k(m) , 0, 0, v(m, "v3-"), 2, 3)
      assertStatus(resp, Success)
      assertEquals(resp.topologyResponse, None)
      assertSuccess(clients.tail.head.get(k(m), 0), v(m, "v3-"))

      val newServer = startClusteredServer(servers.tail.head.getPort + 25)
      try {
         val resp = clients.head.put(k(m) , 0, 0, v(m, "v4-"), 2, 1)
         assertStatus(resp, Success)
         assertEquals(resp.topologyResponse.get.topologyId, currentServerTopologyId)
         val topoResp = resp.asTopologyAwareResponse
         assertEquals(topoResp.members.size, 3)
         (newServer.getAddress :: servers.map(_.getAddress)).foreach(
            addr => assertTrue(topoResp.members.exists(_ == addr)))
         assertSuccess(clients.tail.head.get(k(m), 0), v(m, "v4-"))
      } finally {
         stopClusteredServer(newServer)
         TestingUtil.waitForRehashToComplete(cache(0, cacheName), cache(1, cacheName))
      }

      resp = clients.head.put(k(m) , 0, 0, v(m, "v5-"), 2, 2)
      assertStatus(resp, Success)
      assertEquals(resp.topologyResponse.get.topologyId, currentServerTopologyId)
      var topoResp = resp.asTopologyAwareResponse
      assertEquals(topoResp.members.size, 2)
      servers.map(_.getAddress).foreach(
         addr => assertTrue(topoResp.members.exists(_ == addr)))
      assertSuccess(clients.tail.head.get(k(m), 0), v(m, "v5-"))

      val crashingServer = startClusteredServer(
            servers.tail.head.getPort + 25, doCrash = true)
      try {
         val resp = clients.head.put(k(m) , 0, 0, v(m, "v6-"), 2, 3)
         assertStatus(resp, Success)
         assertEquals(resp.topologyResponse.get.topologyId, currentServerTopologyId)
         val topoResp = resp.asTopologyAwareResponse
         assertEquals(topoResp.members.size, 3)
         (crashingServer.getAddress :: servers.map(_.getAddress)).foreach(
            addr => assertTrue(topoResp.members.exists(_ == addr)))
         assertSuccess(clients.tail.head.get(k(m), 0), v(m, "v6-"))
      } finally {
         stopClusteredServer(crashingServer)
         TestingUtil.waitForRehashToComplete(cache(0, cacheName), cache(1, cacheName))
      }

      resp = clients.head.put(k(m) , 0, 0, v(m, "v7-"), 2, 4)
      assertStatus(resp, Success)
      assertEquals(resp.topologyResponse.get.topologyId, currentServerTopologyId)
      topoResp = resp.asTopologyAwareResponse
      assertEquals(topoResp.members.size, 2)
      servers.map(_.getAddress).foreach(
         addr => assertTrue(topoResp.members.exists(_ == addr)))
      assertSuccess(clients.tail.head.get(k(m), 0), v(m, "v7-"))

      resp = clients.head.put(k(m) , 0, 0, v(m, "v8-"), 3, 1)
      assertStatus(resp, Success)

      checkTopologyReceived(resp.topologyResponse.get, servers, cacheName)
      assertSuccess(clients.tail.head.get(k(m), 0), v(m, "v8-"))
   }

   @Test(enabled=false) // Disable explicitly to avoid TestNG thinking this is a test!!
   protected def checkTopologyReceived(topoResp: AbstractTestTopologyAwareResponse,
           servers: List[HotRodServer], cacheName: String) {
      assertNoHashTopologyReceived(topoResp, servers, cacheName, currentServerTopologyId)
   }

}

File: server/hotrod/src/test/java/org/infinispan/server/hotrod/HotRod10ReplicationTest.java
package org.infinispan.server.hotrod;

import static org.infinispan.server.hotrod.Constants.INTELLIGENCE_BASIC;
import static org.infinispan.server.hotrod.Constants.INTELLIGENCE_HASH_DISTRIBUTION_AWARE;
import static org.infinispan.server.hotrod.Constants.INTELLIGENCE_TOPOLOGY_AWARE;
import static org.infinispan.server.hotrod.OperationStatus.OperationNotExecuted;
import static org.infinispan.server.hotrod.OperationStatus.Success;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.assertKeyDoesNotExist;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.assertNoHashTopologyReceived;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.assertStatus;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.assertSuccess;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.assertTopologyReceived;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.hotRodCacheConfiguration;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.k;
import static org.infinispan.server.hotrod.test.HotRodTestingUtil.v;
import static org.testng.AssertJUnit.assertEquals;
import static org.testng.AssertJUnit.assertTrue;

import java.lang.reflect.Method;
import java.util.List;
import java.util.stream.Stream;

import org.infinispan.configuration.cache.CacheMode;
import org.infinispan.configuration.cache.ConfigurationBuilder;
import org.infinispan.server.hotrod.test.AbstractTestTopologyAwareResponse;
import org.infinispan.server.hotrod.test.TestResponse;
import org.infinispan.test.TestingUtil;
import org.infinispan.topology.ClusterCacheStatus;
import org.testng.annotations.Test;

/**
 * Tests Hot Rod instances configured with replication in protocol version 1.0.
 *
 * @author Galder Zamarreño
 * @since 4.1
 */
@Test(groups = "functional", testName = "server.hotrod.HotRod10ReplicationTest")
public class HotRod10ReplicationTest extends HotRodMultiNodeTest {

   @Override
   protected String cacheName() {
      return "hotRodReplSync";
   }

   @Override
   protected ConfigurationBuilder createCacheConfig() {
      ConfigurationBuilder config = hotRodCacheConfiguration(
            getDefaultClusteredCacheConfig(CacheMode.REPL_SYNC, false));
      config.clustering().stateTransfer().fetchInMemoryState(true);
      return config;
   }

   @Override
   protected byte protocolVersion() {
      return 10;
   }

   public void testReplicatedPut(Method m) {
      TestResponse resp = clients().get(0).put(k(m), 0, 0, v(m));
      assertStatus(resp, Success);
      assertSuccess(clients().get(1).get(k(m), 0), v(m));
   }

   public void testReplicatedPutIfAbsent(Method m) {
      assertKeyDoesNotExist(clients().get(0).assertGet(m));
      assertKeyDoesNotExist(clients().get(1).assertGet(m));
      TestResponse resp = clients().get(0).putIfAbsent(k(m), 0, 0, v(m));
      assertStatus(resp, Success);
      assertSuccess(clients().get(1).get(k(m), 0), v(m));
      assertStatus(clients().get(1).putIfAbsent(k(m), 0, 0, v(m, "v2-")), OperationNotExecuted);
   }

   public void testReplicatedReplace(Method m) {
      TestResponse resp = clients().get(0).replace(k(m), 0, 0, v(m));
      assertStatus(resp, OperationNotExecuted);
      resp = clients().get(1).replace(k(m), 0, 0, v(m));
      assertStatus(resp, OperationNotExecuted);
      clients().get(1).assertPut(m);
      resp = clients().get(1).replace(k(m), 0, 0, v(m, "v1-"));
      assertStatus(resp, Success);
      assertSuccess(clients().get(0).assertGet(m), v(m, "v1-"));
      resp = clients().get(0).replace(k(m), 0, 0, v(m, "v2-"));
      assertStatus(resp, Success);
      assertSuccess(clients().get(1).assertGet(m), v(m, "v2-"));
   }

   public void testPingWithTopologyAwareClient() {
      TestResponse resp = clients().get(0).ping();
      assertStatus(resp, Success);
      assertEquals(resp.topologyResponse, null);

      resp = clients().get(1).ping(INTELLIGENCE_BASIC, 0);
      assertStatus(resp, Success);
      assertEquals(resp.topologyResponse, null);

      resp = clients().get(0).ping(INTELLIGENCE_TOPOLOGY_AWARE, 0);
      assertStatus(resp, Success);
      assertTopologyReceived(resp.topologyResponse, servers(), currentServerTopologyId());

      resp = clients().get(1).ping(INTELLIGENCE_TOPOLOGY_AWARE, 0);
      assertStatus(resp, Success);
      assertTopologyReceived(resp.topologyResponse, servers(), currentServerTopologyId());

      resp = clients().get(1).ping(INTELLIGENCE_TOPOLOGY_AWARE, ClusterCacheStatus.INITIAL_TOPOLOGY_ID + 2 * nodeCount());
      assertStatus(resp, Success);
      assertEquals(resp.topologyResponse, null);
   }

   public void testReplicatedPutWithTopologyChanges(Method m) {
      TestResponse resp = clients().get(0).put(k(m), 0, 0, v(m), INTELLIGENCE_BASIC, 0);
      assertStatus(resp, Success);
      assertEquals(resp.topologyResponse, null);
      assertSuccess(clients().get(1).get(k(m), 0), v(m));

      resp = clients().get(0).put(k(m), 0, 0, v(m, "v1-"), INTELLIGENCE_TOPOLOGY_AWARE, 0);
      assertStatus(resp, Success);
      assertTopologyReceived(resp.topologyResponse, servers(), currentServerTopologyId());

      resp = clients().get(1).put(k(m), 0, 0, v(m, "v2-"), INTELLIGENCE_TOPOLOGY_AWARE, 0);
      assertStatus(resp, Success);
      assertTopologyReceived(resp.topologyResponse, servers(), currentServerTopologyId());

      resp = clients().get(0).put(k(m), 0, 0, v(m, "v3-"), INTELLIGENCE_TOPOLOGY_AWARE,
                                  ClusterCacheStatus.INITIAL_TOPOLOGY_ID + 2 * nodeCount());
      assertStatus(resp, Success);
      assertEquals(resp.topologyResponse, null);
      assertSuccess(clients().get(1).get(k(m), 0), v(m, "v3-"));

      HotRodServer newServer = startClusteredServer(servers().get(1).getPort() + 25);
      try {
         resp = clients().get(0).put(k(m), 0, 0, v(m, "v4-"), INTELLIGENCE_TOPOLOGY_AWARE,
                                     ClusterCacheStatus.INITIAL_TOPOLOGY_ID + 2 * nodeCount());
         assertStatus(resp, Success);
         assertEquals(resp.topologyResponse.topologyId, currentServerTopologyId());
         AbstractTestTopologyAwareResponse topoResp = resp.asTopologyAwareResponse();
         assertEquals(topoResp.members.size(), nodeCount() + 1);
         Stream.concat(Stream.of(newServer), servers().stream())
               .map(HotRodServer::getAddress)
               .forEach(addr -> assertTrue(topoResp.members.contains(addr)));
         assertSuccess(clients().get(1).get(k(m), 0), v(m, "v4-"));
      } finally {
         stopClusteredServer(newServer);
         TestingUtil.waitForNoRebalance(cache(0, cacheName()), cache(1, cacheName()));
      }

      resp = clients().get(0).put(k(m), 0, 0, v(m, "v5-"), INTELLIGENCE_TOPOLOGY_AWARE,
                                  ClusterCacheStatus.INITIAL_TOPOLOGY_ID + 2 * nodeCount() + 2);
      assertStatus(resp, Success);
      assertEquals(resp.topologyResponse.topologyId, currentServerTopologyId());
      AbstractTestTopologyAwareResponse topoResp1 = resp.asTopologyAwareResponse();
      assertEquals(topoResp1.members.size(), nodeCount());
      servers().stream()
               .map(HotRodServer::getAddress)
               .forEach(addr -> assertTrue(topoResp1.members.contains(addr)));
      assertSuccess(clients().get(1).get(k(m), 0), v(m, "v5-"));

      HotRodServer crashingServer = startClusteredServer(servers().get(1).getPort() + 25);
      try {
         resp = clients().get(0).put(k(m), 0, 0, v(m, "v6-"), INTELLIGENCE_TOPOLOGY_AWARE,
                                     ClusterCacheStatus.INITIAL_TOPOLOGY_ID + 2 * nodeCount() + 4);
         assertStatus(resp, Success);
         assertEquals(resp.topologyResponse.topologyId, currentServerTopologyId());
         AbstractTestTopologyAwareResponse topoResp2 = resp.asTopologyAwareResponse();
         assertEquals(topoResp2.members.size(), nodeCount() + 1);
         Stream.concat(Stream.of(crashingServer), servers().stream())
               .map(HotRodServer::getAddress)
               .forEach(addr -> assertTrue(topoResp2.members.contains(addr)));
         assertSuccess(clients().get(1).get(k(m), 0), v(m, "v6-"));
      } finally {
         stopClusteredServer(crashingServer);
         TestingUtil.waitForNoRebalance(cache(0, cacheName()), cache(1, cacheName()));
      }

      resp = clients().get(0).put(k(m), 0, 0, v(m, "v7-"), INTELLIGENCE_TOPOLOGY_AWARE,
                                  ClusterCacheStatus.INITIAL_TOPOLOGY_ID + 2 * nodeCount() + 6);
      assertStatus(resp, Success);
      assertEquals(resp.topologyResponse.topologyId, currentServerTopologyId());
      AbstractTestTopologyAwareResponse topoResp3 = resp.asTopologyAwareResponse();
      assertEquals(topoResp3.members.size(), nodeCount());
      servers().stream()
               .map(HotRodServer::getAddress)
               .forEach(addr -> assertTrue(topoResp3.members.contains(addr)));
      assertSuccess(clients().get(1).get(k(m), 0), v(m, "v7-"));

      resp = clients().get(0).put(k(m), 0, 0, v(m, "v8-"), INTELLIGENCE_HASH_DISTRIBUTION_AWARE,
                                  ClusterCacheStatus.INITIAL_TOPOLOGY_ID + nodeCount());
      assertStatus(resp, Success);

      checkTopologyReceived(resp.topologyResponse, servers(), cacheName());
      assertSuccess(clients().get(1).get(k(m), 0), v(m, "v8-"));
   }

   protected void checkTopologyReceived(AbstractTestTopologyAwareResponse topoResp,
                                        List<HotRodServer> servers, String cacheName) {
      assertNoHashTopologyReceived(topoResp, servers, cacheName, currentServerTopologyId());
   }

}
